### PR TITLE
GEODE-9834: SRANDMEMBER Command Support

### PR DESCRIPTION
**Pull request by Kris-10-0**
### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [NA] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/set/SRandMemberExecutor.java [hunk 1]
```diff
@@ -14,55 +14,13 @@
...
-        set -> set.srandmember(count));
-
-    if (countSpecified) {
-      return RedisResponse.array(results, true);
-    } else if (results.isEmpty()) {
-      return RedisResponse.nil();
-    } else {
-      return RedisResponse.bulkString(results.iterator().next());
-    }
+  protected List<byte[]> preformCommand(RedisSet set, int count) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/set/SRandMemberExecutor.java:23]: is the name a typo? I think it should be "perform" instead of "preform"

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 2]
```diff
@@ -161,49 +162,42 @@ public RedisSet() {}
...
+    if (!duplicatesAllowed && membersSize <= count) {
+      randMembers.addAll(members);
     } else {
-      Set<byte[]> result = new MemberSet(count);
-      // Note that rand.nextInt can return duplicates when "count" is high
-      // so we need to use a Set to collect the results.
-      while (result.size() < count) {
-        byte[] s = entries[rand.nextInt(entries.length)];
-        result.add(s);
+      Set<Integer> usedPos = null;
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:176]: would "usedIndexes" be a better name?

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 3]
```diff
@@ -161,49 +162,42 @@ public RedisSet() {}
...
+      if (!duplicatesAllowed) {
+        usedPos = new HashSet<>(count);
+      }
+
+      for (int i = 0; i < count; i++) {
+        int randNum = rand.nextInt(memberMapSize);
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:182]: would "randomIndex" be a better name?

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 4]
```diff
@@ -161,49 +162,42 @@ public RedisSet() {}
...
+        while (member == null || (!duplicatesAllowed && usedPos.contains(randNum))) {
+          // TODO: Can a member be null?
+          if (!duplicatesAllowed && member == null) {
+            usedPos.add(randNum);
+          }
+          randNum = rand.nextInt(memberMapSize);
+          member = members.getKeyAtIndex(randNum);
+        }
+
+        randMembers.add(members.getKeyAtIndex(randNum));
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:194]: I think this can be simplified to "randMembers.add(member);"

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]
Nice job using an index on members instead of making a copy of the entire members set!

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 5]
```diff
@@ -161,49 +162,42 @@ public RedisSet() {}
...
+        usedPos = new HashSet<>(count);
+      }
+
+      for (int i = 0; i < count; i++) {
+        int randNum = rand.nextInt(memberMapSize);
+        byte[] member = members.getKeyAtIndex(randNum);
+
+        while (member == null || (!duplicatesAllowed && usedPos.contains(randNum))) {
+          // TODO: Can a member be null?
+          if (!duplicatesAllowed && member == null) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:187]: I think you can remove this "if" block. If member is null we don't add it to the result. By adding the randNum that is an index to "null" you make the usedPos set bigger but if in the future we get the same value in "randNum" we still also call "getKeyAtIndex" and it will return "null". So in the while loop we see "member == null" is true so we don't even do the usedPos.contains check. So I think it would be better to get rid of this "if" block and only add to usedPos when we actually add to "randMembers" which we already do down on line 195.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 6]
```diff
@@ -161,49 +162,42 @@ public RedisSet() {}
...
-      return result;
-    }
-    if (duplicatesAllowed) {
-      List<byte[]> result = new ArrayList<>(count);
-      while (count > 0) {
-        result.add(entries[rand.nextInt(entries.length)]);
-        count--;
-      }
-      return result;
+    List<byte[]> randMembers = new ArrayList<>(count);
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:172]: would "randomMembers" be a better name? Or maybe even better "result"?

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]

**Review by dschneider-pivotal — APPROVED** [review]

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 7]
```diff
@@ -161,49 +162,38 @@ public RedisSet() {}
...
+      Set<Integer> usedIndexes = null;
+      if (!duplicatesAllowed) {
+        usedIndexes = new HashSet<>(count);
+      }
+
+      for (int i = 0; i < count; i++) {
+        int randIndex = rand.nextInt(memberMapSize);
+        byte[] member = members.getKeyAtIndex(randIndex);
+
+        while (member == null || (!duplicatesAllowed && usedIndexes.contains(randIndex))) {
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:185]: This loop could end up taking a long time to execute if `members` is large, duplicates are not allowed and `count` is close to `members.size()`, as it requires generating a random number that isn't the same as any of the previously generated numbers. Including the outer for loop, this ends up making the srandmember command execute in O(N^2) (or thereabouts) which is not in line with native Redis' quoted complexity of O(N).

For the no duplicates case, native Redis uses a different implementation where, if count is greater than 1/3 of the size of the member set, they instead create a duplicate of the entire set and then remove members from it at random until it has the correct size.  From t_set.c (SRANDMEMBER_SUB_STRATEGY_MUL is defined as 3):
>    /* CASE 3:
     * The number of elements inside the set is not greater than
     * SRANDMEMBER_SUB_STRATEGY_MUL times the number of requested elements.
     * In this case we create a set from scratch with all the elements, and
     * subtract random elements to reach the requested number of elements.
     *
     * This is done because if the number of requested elements is just
     * a bit less than the number of elements in the set, the natural approach
     * used into CASE 4 is highly inefficient. */

It's possible that there's a better approach than this, but it's an improvement over the current one, at least.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  @Test
+  public void srandmemberWithCount_withExistentSet_returnsEmptySet() {
+    jedis.sadd(setKey, setMembers);
+    int count = 2;
+
+    List<String> result = jedis.srandmember(setKey, count);
+    assertThat(result.size()).isEqualTo(2);
+    assertThat(result).containsAnyElementsOf(Arrays.asList(setMembers));
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:100]: This assertion is only testing that `result` contains at least one element found in `setMembers`. The correct assertion to use here would be `isSubsetOf(setMembers)`, along with `assertThat(result).doesNotHaveDuplicates()` to ensure that only unique members are returned.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  @Test
+  public void srandmemberWithoutCount_withExistentSet_returnsOneMember() {
+    jedis.sadd(setKey, setMembers);
+
+    String result = jedis.srandmember(setKey);
+    assertThat(setMembers).contains(result);
+  }
+
+  @Test
+  public void srandmemberWithCount_withExistentSet_returnsEmptySet() {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:94]: This test name is incorrect. It should probably be "srandmemberWithCount_withExistentSet_returnsCorrectNumberOfMembers".

It would also be good to have a test showing that when count is positive and larger than the size of the set, no duplicate entries are returned.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    jedis.sadd(setKey, setMembers);
+    int count = setMembers.length;
+
+    List<String> result = jedis.srandmember(setKey, count);
+    assertThat(result.size()).isEqualTo(count);
+    assertThat(result).containsExactlyInAnyOrder(setMembers);
+  }
+
+  @Test
+  public void srandmemberWithCountAsGreaterThanSetSize_withExistentSet_returnsAllMembersWithDuplicates() {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:114]: This test name is a little misleading, as the test is actually testing a negative count value, not a count that is greater than the set size.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    jedis.sadd(setKey, setMembers);
+    int count = -20;
+
+    List<String> result = jedis.srandmember(setKey, count);
+    assertThat(result.size()).isEqualTo(-count);
+    for (String s : result) {
+      assertThat(s).isNotNull();
+      assertThat(setMembers).contains(s);
+    }
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:123]: This can be simplified to:
```
assertThat(result).isSubsetOf(setMembers);
```

**Review by DonalEvans — CHANGES_REQUESTED** [review]

**Comment by DonalEvans** [comment]
In addition to the comments I left, `AbstractSetsIntegrationTest` should be modified to remove tests related to srandmember, since those tests now belong in `AbstractSRandMemberIntegrationTest`. `AbstractSetsIntegrationTest` should probably be renamed to `AbstractSAddIntegrationTest` and all non-sadd tests moved out of it, really, but maybe that's better left to another PR.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/set/SetRandomExecutor.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.apache.geode.redis.internal.commands.Command;
+import org.apache.geode.redis.internal.commands.executor.CommandExecutor;
+import org.apache.geode.redis.internal.commands.executor.RedisResponse;
+import org.apache.geode.redis.internal.data.RedisKey;
+import org.apache.geode.redis.internal.data.RedisSet;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+import org.apache.geode.redis.internal.services.RegionProvider;
+
+public abstract class SetRandomExecutor implements CommandExecutor {
```
[the remark this answers is not in the document]
Comment (Kris-10-0) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/set/SetRandomExecutor.java:33]: I was going to use it for SPOP, which is basically SRANDMEMBER, but it removes the members.

[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/set/SetRandomExecutor.java:33]: Ah, good call. I hadn't thought about SPOP being random.

**Comment by Kris-10-0** [comment]
Forced push to rebase

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSetsIntegrationTest.java [hunk 13]
```diff
@@ -98,67 +97,6 @@ public void testSAdd_canStoreBinaryData() {
...
-    assertThatThrownBy(() -> jedis.sendCommand("key", Protocol.Command.SMEMBERS))
-        .hasMessageContaining("ERR wrong number of arguments for 'smembers' command");
-  }
-
-  @Test
-  public void smembers_givenMoreThanTwoArguments_returnsWrongNumberOfArgumentsError() {
-    assertThatThrownBy(() -> jedis
-        .sendCommand("key", Protocol.Command.SMEMBERS, "key", "extraArg"))
-            .hasMessageContaining("ERR wrong number of arguments for 'smembers' command");
-  }
```
Comment (nonbinaryprogrammer) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSetsIntegrationTest.java:160]: did you mean to delete these smembers ones too?

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 14]
```diff
@@ -202,48 +202,50 @@ static int setOpStoreResult(RegionProvider regionProvider, RedisKey destinationK
...
+
+  private void srandomUniqueList(int count, int memberMapSize, List<byte[]> result) {
+    List<Integer> allIndexes = new ArrayList<>(memberMapSize);
+    for (int i = 0; i < memberMapSize; i++) {
+      allIndexes.add(i);
+    }
+    Collections.shuffle(allIndexes);
+    int i = 0;
+    while (result.size() != count) {
+      byte[] member = members.getKeyAtIndex(i);
```
Comment (nonbinaryprogrammer) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:243]: why not just call `getKey(pos)` directly?

**Review by nonbinaryprogrammer — COMMENTED** [review]
just a couple of questions

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java (hunk 14 above)
Comment (Kris-10-0) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java:243]: I think I was worried since getKey(pos) returned an object K it wouldn't be a byte[] when I called it directly. I tried the direct call and it worked. So I'll make the fix.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSetsIntegrationTest.java (hunk 13 above)
Comment (Kris-10-0) [reply geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSetsIntegrationTest.java:160]: Nope accident.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java [hunk 15]
```diff
@@ -161,49 +162,38 @@ public RedisSet() {}
...
-    }
-
-    if (!duplicatesAllowed && membersSize <= count && count != 1) {
-      return new ArrayList<>(members);
-    }
+    count = duplicatesAllowed ? -count : count;
+    int membersSize = members.size();
+    int memberMapSize = members.getMemberMapSize();
 
     Random rand = new Random();
```
[the remark this answers is not in the document]
Comment (Kris-10-0) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java]: @DonalEvans Creating a final field for the class messes with the size of the bytes causing tests in RedisSet to fail.

[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java]: That's fine, those tests (and probably the value defined in `expectedPerEntryOverhead()` in MemoryOverheadIntegrationTest) can be adjusted to account for the new size.

**Review by nonbinaryprogrammer — APPROVED** [review]

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/collections/SizeableObjectOpenCustomHashSet.java [hunk 16]
```diff
@@ -62,6 +62,14 @@ public boolean remove(Object k) {
     return removed;
   }
 
+  public K getKey(final int pos) {
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/collections/SizeableObjectOpenCustomHashSet.java:65]: This method name would be more accurately "getElementForIndex()" or "getFromBackingArray()" or something similar, since a Set doesn't really have a concept of keys.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

**Comment by Kris-10-0** [comment]
Forced push to rebase

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Protocol;
+
+import org.apache.geode.redis.RedisIntegrationTest;
+
+public abstract class AbstractSRandMemberIntegrationTest implements RedisIntegrationTest {
+  private JedisCluster jedis;
+  private static final String NON_EXISTENT_SET_KEY = "{user1}nonExistentSet";
+  private static final String SET_KEY = "{user1}setKey";
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:41]: Sorry I didn't catch this earlier, but could these tags be `{tag1}` to follow the new convention?

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java [hunk 18]
```diff
@@ -64,84 +66,100 @@ public void srandmemberTooManyArgs_returnsError() {
...
+  public void srandmember_withoutCount_withExistentSet_returnsOneMember() {
+    jedis.sadd(SET_KEY, SET_MEMBERS);
+
+    String result = jedis.srandmember(SET_KEY);
+    assertThat(result).isIn(Arrays.asList(SET_MEMBERS));
   }
 
   @Test
-  public void srandmemberWithoutCount_withExistentSet_returnsOneMember() {
+  public void srandmember_withCountAsZero_withExistentSet_returnsSubsetOfSet() {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/set/AbstractSRandMemberIntegrationTest.java:89]: This test name is inaccurate. It seems like it should be "srandmember_withCountAsZero_withExistentSet_returnsEmptyList"

**Review by DonalEvans — CHANGES_REQUESTED** [review]

**Review by DonalEvans — APPROVED** [review]
